### PR TITLE
Add 'last' functionality: limit returned tasks from things.database.Database.get_tasks() to 'last' X days or weeks.

### PR DESCRIPTION
**Pull request by lmgibson**
Hello! This pull request adds the parameter 'last' to `things.database.Database.get_tasks()` which is an additional filter on the SQL query so that the returned tasks are only tasks from the last X days or weeks. 

The functionality is implemented by adding a new line to the `where_predicate` string that filters the tasks based on the user-specified number of days, or weeks, from the current date. The function that formats the query is `make_date_filter()` which is a series of if / else statements to ensure the validity of the value given to 'last'. If no value is given to 'last' then `make_date_filter()` will return an empty string.

In addition to adding the functionality, this pull request adds a basic test that checks if providing a valid input to 'last' will return an empty list without error. Note, this test assumes that the test database will never change and therefore no new tasks are created that may cause this test to fail.

Happy to discuss these changes in more detail or make any changes in case I missed something. Thank you!

### CODE REVIEW

File: things/database.py [hunk 1]
```diff
@@ -683,6 +685,30 @@ def make_filter(column, value):
...
+    Args:
+        create_var_name (string): Name of the column that has information on when each task was created.
+        last (string): A string comprised of an integer and a single character that can be 'd' or 'w' that determines whether to return all tasks for the past X days or weeks.
+
+    Returns:
+        string: A string to add into the filter for the SQL query.
+    """
+    if (type(last) is str):
+        if last[-1] == "d":
+            return f'AND datetime(TASK.{create_var_name}, "unixepoch", "localtime") BETWEEN datetime("now", "-{int(last[:-1])} days") AND datetime("now", "localtime")'
```
Comment (mikez) [inline things/database.py:700]: You might simplify to:
```python
number = int(offset[:-1])

if suffix == "d":
    modifier = f"-{number} days"
elif suffix == "w":
    modifier = f"-{7 * number} days"
```
...

return sql filter:
```sql
datetime(TASK.{created_column}, "unixepoch", "localtime") > datetime("now", "{modifier}")
```

**Review by mikez — CHANGES_REQUESTED** [review]
@lmgibson Good job! Added some comments related to keeping a consistent style. What do you think?

File: things/api.py [hunk 2]
```diff
@@ -111,6 +111,9 @@ def tasks(uuid=None, include_items=False, **kwargs):  # noqa: C901
     count_only : boolean, default False
         Only output length of result. This is done by a SQL COUNT query.
 
+    last : str, default None, optional
+        Limit returned tasks to tasks created in the last X days or weeks. Takes as input string of format 'X[d/w]' where X is a non-negative integer and 'd' stands for days and 'w' for weeks.
```
[the remark this answers is not in the document]
Comment (lmgibson) [reply things/api.py]: Reduced the length of the docstring in the most recent commit. Removed everything past the first sentence.

File: things/database.py [hunk 3]
```diff
@@ -683,6 +685,30 @@ def make_filter(column, value):
...
 
+def make_date_filter(create_var_name, last):
+    """Returns SQL filter to limit date range of the SQL query.
+
+    Args:
+        create_var_name (string): Name of the column that has information on when each task was created.
+        last (string): A string comprised of an integer and a single character that can be 'd' or 'w' that determines whether to return all tasks for the past X days or weeks.
+
+    Returns:
+        string: A string to add into the filter for the SQL query.
```
[the remark this answers is not in the document]
Comment (lmgibson) [reply things/database.py]: See most recent commit for reduced docstring. Let me know if it looks okay!

File: things/database.py [hunk 4]
```diff
@@ -683,6 +685,30 @@ def make_filter(column, value):
...
+    """
+    if (type(last) is str):
```
[the remark this answers is not in the document]
Comment (lmgibson) [reply things/database.py]: Thank you! This was very helpful advice. In the most recent commit, I converted last to offset, added the proposed validation steps along with an additional one that handles the case where offset = None (not specified).

File: things/database.py (hunk 1 above)
Comment (lmgibson) [reply things/database.py]: Great suggestions. I added them in the most recent commit.

**Comment by lmgibson** [comment]
New commit that incorporates suggested changes by @mikez. The modifications are such:

1. Reduced length of docstrings for `things.database.make_date_filter()` and `things.api.tasks()`
2. Separated validation from the meat of `things.database.make_date_filter()`
3. In `things.database.make_date_filter` added handling for when last is not specified and defaults to None.
4. Chunked `things.database.make_date_filter()` to reduce the complexity of the function.

Thanks again for the feedback. Let me know what you think!

**Comment by AlexanderWillner** [comment]
Minor issue: `things/database.py:712:8: E713 test for membership should be 'not in'`

**Comment by lmgibson** [comment]
Thanks, @AlexanderWillner for pointing that out. Addressed it with a minor change in the most recent commit.

**Comment by lmgibson** [comment]
> Minor issue: `things/database.py:712:8: E713 test for membership should be 'not in'`

Thanks, @AlexanderWillner for pointing that out. Addressed it with a minor change in the most recent commit.

**Comment by AlexanderWillner** [comment]
Great. Also have a short look at https://github.com/thingsapi/things.py/pull/52/checks?check_run_id=2326339753 / run `make lint` locally.

File: things/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikez) [reply things/api.py]: @lmgibson To clarify: multiple lines are wonderful! I think the content of your docstring was great. My intented suggestion was to wrap it into multiple lines of at most at 80-88 characters.

File: things/api.py [hunk 5]
```diff
@@ -112,7 +112,7 @@ def tasks(uuid=None, include_items=False, **kwargs):  # noqa: C901
         Only output length of result. This is done by a SQL COUNT query.
 
     last : str, default None, optional
-        Limit returned tasks to tasks created in the last X days or weeks. Takes as input string of format 'X[d/w]' where X is a non-negative integer and 'd' stands for days and 'w' for weeks.
+        Limit returned tasks to tasks created in the last X days or weeks.
```
Comment (mikez) [inline things/api.py:115]: My intended suggestion was to wrap into multiple lines of at most 80-88 characters each. Your content was great! :)

File: things/database.py [hunk 6]
```diff
@@ -685,28 +685,38 @@ def make_filter(column, value):
     }.get(value, default)
 
 
-def make_date_filter(create_var_name, last):
-    """Returns SQL filter to limit date range of the SQL query.
+def make_date_filter(created_column, offset):
+    """
+    Returns SQL filter to limit date range of the SQL query.
+    """
```
Comment (mikez) [inline things/database.py:691]: Please do keep Args and Returns, I like it! My intended suggestion was to rename "Args" into "Parameters" and slightly modify the styling to be consistent with others. See `things.api.tasks` for an example of the numpy/pandas-like docstrings.

File: things/database.py [hunk 7]
```diff
@@ -685,28 +685,38 @@ def make_filter(column, value):
...
-        else:
-            print("Please specify last as a string of the format 'Xd' where X is a non-negative integer and 'd' states that it is measured in days")    
-    else:
-        if last:
-            print("Please specify last as a string of the format 'Xd' where X is a non-negative integer and 'd' states that it is measured in days")    
-        else:
-            return ""
+    if not isinstance(offset, str):
+        raise ValueError(
+            "Please specify last as a string of the format 'X[d/w]' where X is a non-negative integer followed by 'd' or 'w' that indicates days or weeks."
```
Comment (mikez) [inline things/database.py:701]: In Python, you can wrap long strings into multiple lines of up to 88 characters each like so:

```python
raise ValueError(
    "Please specify last as a string of the format 'X[d/w]'"
    "where X is a non-negative integer followed by 'd' or 'w'"
    "that indicates days or weeks."
)
```
It will automatically concatinate the individual strings.

File: things/database.py [hunk 8]
```diff
@@ -685,28 +685,38 @@ def make_filter(column, value):
...
+        )
+
+    suffix = offset[-1]
+    if suffix not in ("d", "w"):
+        raise ValueError(
+            "Please specify last as a string of the format 'X[d/w]' where X is a non-negative integer followed by 'd' or 'w' that indicates days or weeks."
```
Comment (mikez) [inline things/database.py:707]: If you want to output the same error message, I'd wrap the two checks into one like so:

```python
if not isinstance(offset, str) or offset[-1] not in ("d", "w"):
    ValueError(...)
```

Comment (mikez) [reply things/database.py:707]: The philosophy I use here is "don't repeat yourself" or "don't copy paste" :) — It makes maintenance easier down the road when things change.

**Review by mikez — CHANGES_REQUESTED** [review]
@lmgibson Thank you for your patience with this review. 💛  My intention here is not to be nitpicky, but rather consistency and transmitting the joy of Python code that flows nicely. :)

File: things/database.py (hunk 8 above)
Comment (mikez) [reply things/database.py:707]: Also, maybe the error message could refer to the parameter as `offset`?

For example:

```python
ValueError('Invalid offset: {offset!r}.\nA valid offset is a string of …')
```

File: things/database.py (hunk 6 above)
Comment (lmgibson) [reply things/database.py:691]: Oh! Apologies for misinterpreting. I'm glad you liked them. I'll add them back in.

**Comment by lmgibson** [comment]
> @lmgibson Thank you for your patience with this review. 💛 My intention here is not to be nitpicky, but rather nurture consistency and transmit the joy of Python code that flows nicely. :)

Of course! This has been a great learning experience so I don't mind the eye for detail. 

I just pushed a new commit that brings back the old docstrings but formats them to match the other docstrings in the pandas/numpy convention. I also cleaned up the ValueError strings but made the decision to make them more specific and keep them separate. I also used `last` instead of `offset` inside of the ValueError strings because I think that most people will be interacting with offset via the `last` parameter. Although, I just realized that maybe we could simply change `last` to `offset` in case that is a better name.

Let me know what you think! I don't mind if there are more changes.

**Comment by mikez** [comment]
@lmgibson Looks great to me! I'm happy to merge.

If you want to keep going, here are two minor things that we might change:

1. Move validation of offsets into its own function `validate_offset` akin to the function `validate`. You can then use this new `validate_offset` function in the beginning of `get_tasks` to validate `last`. You can also use `last` in the error message.
2. You could rename `created_column` to `date_column` since your function would work for any date columns at this point.

These two things are optional. Let me know what you'd enjoy doing. :-)

**Comment by mikez** [comment]
@lmgibson I'm about to do some refactoring. I'm merging this in now. Feel free to change later again. :)

**Comment by lmgibson** [comment]
> @lmgibson Looks great to me! I'm happy to merge.
> 
> If you want to keep going, here are two minor things that we might change:
> 
> 1. Move validation of offsets into its own function `validate_offset` akin to the function `validate`. You can then use this new `validate_offset` function in the beginning of `get_tasks` to validate `last`. You can also use `last` in the error message.
> 2. You could rename `created_column` to `date_column` since your function would work for any date columns at this point.
> 
> These two things are optional. Let me know what you'd enjoy doing. :-)

Great, thank you! I'll take a look at adding the two suggestions and submitting a new PR.
